flink-runtime/src/main/java/org/apache/flink/runtime/registration/RegisteredRpcConnection.java

```diff
@@ -88,24 +86,17 @@ public void start() {
 		pendingRegistration = checkNotNull(generateRegistration());
 		pendingRegistration.startRegistration();
 
-		Future<Tuple2<Gateway, Success>> future = pendingRegistration.getFuture();
-
-		Future<Void> registrationSuccessFuture = future.thenAcceptAsync(new AcceptFunction<Tuple2<Gateway, Success>>() {
-			@Override
-			public void accept(Tuple2<Gateway, Success> result) {
-				targetGateway = result.f0;
-				onRegistrationSuccess(result.f1);
-			}
-		}, executor);
-
-		// this future should only ever fail if there is a bug, not if the registration is declined
-		registrationSuccessFuture.exceptionallyAsync(new ApplyFunction<Throwable, Void>() {
-			@Override
-			public Void apply(Throwable failure) {
-				onRegistrationFailure(failure);
-				return null;
-			}
-		}, executor);
+		CompletableFuture<Tuple2<Gateway, Success>> future = pendingRegistration.getFuture();
+
+		future.whenCompleteAsync(
+			(Tuple2<Gateway, Success> result, Throwable failure) -> {
+				if (failure != null) {
```

Comment:
I somehow like the failure case treating being first. It something like a termination condition for recursion, you also start with it instead of the recursion step.